// esri-loader-react/lib/index.js
'use strict';

exports.__esModule = true;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _esriLoader = require('esri-loader');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var EsriLoaderReact = function (_PureComponent) {
  _inherits(EsriLoaderReact, _PureComponent);

  function EsriLoaderReact() {
    _classCallCheck(this, EsriLoaderReact);

    return _possibleConstructorReturn(this, _PureComponent.apply(this, arguments));
  }

  EsriLoaderReact.prototype.componentDidCatch = function componentDidCatch(error, info) {
    var onError = this.props.onError;


    if (onError) {
      onError(error, info);
    }
  };

  EsriLoaderReact.prototype.componentDidMount = function componentDidMount() {
    var _this2 = this;

    var _props = this.props,
        modulesToLoad = _props.modulesToLoad,
        options = _props.options,
        onReady = _props.onReady,
        onError = _props.onError;


    (0, _esriLoader.loadModules)(modulesToLoad ? modulesToLoad : [], options).then(function (loadedModules) {

      if (onReady) {
        onReady({ loadedModules: loadedModules, containerNode: _this2.mapContainer });
      }
    }).catch(function (error) {

      if (onError) {
        onError(error, null);
      }
    });
  };

  EsriLoaderReact.prototype.render = function render() {
    var _this3 = this;

    var _props2 = this.props,
        renderMapContainer = _props2.renderMapContainer,
        mapContainerClassName = _props2.mapContainerClassName,
        children = _props2.children;


    if (!renderMapContainer) {
      return children ? children : null;
    }

    return _react2.default.createElement(
      'div',
      { ref: function ref(node) {
          return _this3.mapContainer = node;
        }, className: mapContainerClassName },
      children ? children : null
    );
  };

  return EsriLoaderReact;
}(_react.PureComponent);

EsriLoaderReact.propTypes = process.env.NODE_ENV !== "production" ? {
  renderMapContainer: _propTypes2.default.bool, // default is true
  mapContainerClassName: _propTypes2.default.string, // default is 'map-view'
  modulesToLoad: _propTypes2.default.arrayOf(_propTypes2.default.string),
  options: _propTypes2.default.shape({
    url: _propTypes2.default.string,
    dojoConfig: _propTypes2.default.object
  }),
  onError: _propTypes2.default.func, // (error, info) =>
  onReady: _propTypes2.default.func // ({loadedModules, containerNode (null if renderMapContainer !== true)})
} : {};

EsriLoaderReact.defaultProps = {
  renderMapContainer: true,
  mapContainerClassName: 'map-view',
  onError: function onError(error, info) {
    return console.error(error);
  }
};

exports.default = EsriLoaderReact;
module.exports = exports['default'];
